DOC: Enumerate the differences between numpy and numpy.array_api

asmeurer opened the pull request:
This is a more organized presentation of the various "Note" comments in numpy/array_api.

In particular, each difference is notated as to whether it is a strictness difference (no change in NumPy needed), a compatible change (NumPy can change in a backwards compatible way), or a breaking change (NumPy would need to break backwards compatibility to match the spec).

This is related to the discussion in https://github.com/numpy/numpy/issues/21135. I'm hopeful that this can help the NumPy developers better understand exactly how NumPy would need to change in order for the top-level `numpy` namespace to become array API compliant. 

I've also included a few minor fixes to numpy.array_api. 

Note that, as outlined in the original pull request (https://github.com/numpy/numpy/pull/18585), this does not include any documentation for the array API itself. It would be a good idea to do this, but it is a harder problem, because each function would need a docstring. For now, the [spec](https://data-apis.org/array-api/latest/index.html) serves as the best reference for `numpy.array_api`.

<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      http://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      http://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->
…and vecdot

asmeurer committed:
Update some Note comments in numpy.array_api

asmeurer committed:
Remove some outdated comments

asmeurer committed:
Properly restrict the input dtypes for the array_api trace, svdvals, and vecdot

asmeurer committed:
Add some missing notes to array_api

asmeurer committed:
Fix a type hint in numpy.array_api

asmeurer committed:
Add a document that enumerates the differences between numpy and numpy.array_api

This is a more organized presentation of the various "Note" comments in
numpy/array_api.

In particular, each difference is notated as to whether it is a strictness
difference (no change in NumPy needed), a compatible change (NumPy can change
in a backwards compatible way), or a breaking change (NumPy would need to
break backwards compatibility to match the spec).

rgommers commented on doc/source/reference/array_api.rst:
```diff
@@ -0,0 +1,787 @@
+     - **Compatible**
+     -
+   * - ``unique_all`` and ``unique_indices`` return indices with the same
+       shape as ``x``.
```
Given that these are new function names, the `breaking` label seems wrong - this is `compatible`. It's good to point out separately as a note so fine to keep, but it's backwards compatible.

asmeurer replied:
Ah that's a good point. I made that same observation for some other new functions, but missed it here.

rgommers commented on doc/source/reference/array_api.rst:
```diff
@@ -0,0 +1,787 @@
+.. _array_api:
+
+*********
+Array API
```
I suggest calling this "Array API Standard compatibility"

rgommers commented on doc/source/reference/array_api.rst:
```diff
@@ -0,0 +1,787 @@
+*********
+Array API
+*********
+
```
Let's add a `.. note::` here about the implementation in NumPy currently being experimental.

rgommers reviewed (commented):
This looks great and is super useful, thanks @asmeurer! I have only a couple of small comments about the document.

Given that this is very useful, I recommend merging it very soon. Any discussion on making code changes to NumPy should be done elsewhere. Let's focus the review here on whether the document is correct, well-written and useful to have in the NumPy html docs.

rgommers commented:
For some reason the CI link to the artifact is broken. Here is the current rendered version: https://output.circle-artifacts.com/output/job/d6f9451b-164d-4a89-ad56-024550656ddf/artifacts/0/doc/build/html/reference/array_api.html

rgommers commented on numpy/array_api/linalg.py:
```diff
@@ -380,7 +384,7 @@ def vecdot(x1: Array, x2: Array, /, *, axis: int = -1) -> Array:
 # The type for ord should be Optional[Union[int, float, Literal[np.inf,
 # -np.inf]]] but Literal does not support floating-point literals.
-def vector_norm(x: Array, /, *, axis: Optional[Union[int, Tuple[int, int]]] = None, keepdims: bool = False, ord: Optional[Union[int, float]] = 2) -> Array:
+def vector_norm(x: Array, /, *, axis: Optional[Union[int, Tuple[int, ...]]] = None, keepdims: bool = False, ord: Optional[Union[int, float]] = 2) -> Array:
```
Please let's not have that conversation again, we decided this in the original PR and this change does not touch line length. This diff is fine.

asmeurer committed:
Address review comments in the array API documentation

asmeurer committed:
Add a note about the array API copy flag to reshape (which is not yet implemented)

seberg reviewed (commented):
Nice! Generally, should probably just merge this as soon as you think it is ready.  It is a great overview to see that most difference are indeed small.  Besides the "big" ones (promotion, scalars), most indeed seem like small changes.
(I am not too interested in the "strictness" changes, but it is probably good to list them together – in a sense we have a dual role of informing about the `array_api` as well as how the main namespace is incompatible to the "Array API".)

There is a quite large amount of "compatible additions" that we probably have to figure out what to do about if we want them all in the main namespace.  Especially if old name need to stay around.
E.g. the `asin`, etc. ones I guess we need to keep the old names around for a very long time if we want to rename them.  But maybe try to hide them away a bit?

(From a NumPy perspective, I would have almost liked `np.bitwise.left_shift` rather than `np.bitwise_left_shift` for example.)

asmeurer commented on doc/source/reference/array_api.rst:
```diff
@@ -0,0 +1,795 @@
+       the behavior with Python scalars).
+   * - No casting unsigned integer dtypes to floating dtypes (e.g., ``int64 +
+       uint64 -> float64``.
+     - **Strictness**
```
Yeah, I was also unsure about this, but other folks here seemed to agree that the spec does not disallow any additional type promotions, including this one (CC @rgommers @kgryte).

rgommers replied:
I think it's not breaking. There is no `uint64 + int64` promotion rule, which means it is undefined behavior. Libraries may decide to raise an exception, cast to `float64`, cast to `int64`, or do whatever else they like.

rgommers replied:
>  I guess that distinction would only mainly matter if there is a chance that you may want to specify contradicting promotions it in the future.

That's indeed the one/main reason to specify particular exceptions or "implementing this is not allowed". We've tried to stay away from specifying exceptions or error handling type things, because it's impossible to be comprehensive (in addition for it to be difficult to align libraries' behavior here).

asmeurer commented on doc/source/reference/array_api.rst:
```diff
@@ -0,0 +1,795 @@
+     - The spec does not have array scalars, only 0-D arrays. It is not clear
+       if NumPy's scalars deviate from the spec 0-D array behavior in any ways
+       other than the ones outlined in
+       :ref:`array_api-type-promotion-differences`.
```
Right, I wasn't clear if there were any other odd differences. Immutability shouldn't be a problem because strictly speaking the spec allows it (some libraries like Jax are completely immutable, c.f. https://data-apis.org/array-api/latest/design_topics/copies_views_and_mutation.html).

asmeurer replied:
I changed this to "strictness" for now. I believe that other than type promotion (which is discussed elsewhere), scalars can duck type as 0-D arrays sufficiently for the spec. If this is ever disproved, we can update this. It's hard to know for sure without plugging them into the test suite somehow, unless you are aware of any other differences that could affect things.

asmeurer commented on doc/source/reference/array_api.rst:
```diff
@@ -0,0 +1,795 @@
+     - ???
+     -
+   * - ``diagonal`` operates on the last two axes.
+     - **Breaking**
```
I hadn't considered that. I'll leave it as "breaking" but add a note about this.

This is definitely one of the worst "breaking" cases here for NumPy. NumPy's `diagonal` is completely spec noncompliant, and not in a small way either, since the semantics completely differ when you have rank > 2. I don't know if the current behavior is deprecatable either.

asmeurer commented on doc/source/reference/array_api.rst:
```diff
@@ -0,0 +1,795 @@
+       two axes. See `the spec definition
+       <https://data-apis.org/array-api/latest/API_specification/generated/signatures.linear_algebra_functions.matrix_transpose.html#signatures.linear_algebra_functions.matrix_transpose>`__
+   * - ``outer`` only supports 1-dimensional arrays.
+     - **Breaking**
```
@kgryte and I were discussing this internally. It was his view that it is incompatible, but maybe he can comment here.

kgryte replied:
Current NumPy behavior flattens arrays with more than 1 dimension. We should consider this incompatible, especially as the consortium is likely to move forward with support for batching: https://github.com/data-apis/array-api/issues/242. If 242 moves forward, NumPy's behavior would be non-compliant. And as such, I think it is appropriate here to indicate that the spec entails a breaking change.

seberg replied:
Thanks, I didn't realize `np.outer` used raveling. `np.ufunc.outer` result is `a.shape + b.shape` I think.  Adding an `axis` argument and changing its default might work, but then breaking is definitely right for now anyway.

asmeurer commented on doc/source/reference/array_api.rst:
```diff
@@ -0,0 +1,795 @@
+     - Notes
+   * - ``sum`` and ``prod`` always upcast ``float32`` to ``float64`` when
+       ``dtype=None``.
+     - **Breaking**
```
Yes, there's [no restriction on axis](https://data-apis.org/array-api/latest/API_specification/generated/signatures.statistical_functions.sum.html). It's not clear to me that the axis should matter? It wasn't ever mentioned in [the discussion](https://github.com/data-apis/array-api/issues/202) as far as I can tell.

seberg replied:
I faintly remember seeing the discussion and being surprised by it before, but not diving into it and thinking seriously about it.  I feel this change may be something to run by a wider audience.  I guess it may not be massive from a back-compat point of view if it only affects float16, float32, and complex64.

Also: I think this is at least the default float/integer precision when the input is float/integer?  Or do integer inputs return floats?

kgryte replied:
The default output dtype is based on the input dtype: [spec](https://github.com/data-apis/array-api/blob/c58051748090fe97aae980837c63c02c5e405ff2/spec/API_specification/signatures/statistical_functions.py#L111).

If the input is an integer dtype, the default output dtype is the default integer dtype provided the default integer dtype has sufficient precision.

Similarly, if the input is a float dtype, the default output dtype is the default floating-point dtype, provided the default floating-point dtype has sufficient precision.

In short, when `dtype=None`, the output array should never have a dtype with less precision than the input dtype.

asmeurer commented on doc/source/reference/array_api.rst:
```diff
@@ -0,0 +1,795 @@
+     - No cross-kind casting. No value-based casting on scalars.
+   * - ``stack`` has different default casting rules from ``np.stack``
+     - **Strictness**
+     - No cross-kind casting. No value-based casting on scalars.
```
`stack` and `concat` do allow promoting, but cross-kind promoting is unspecified ("strictness"). c.f. https://data-apis.org/array-api/latest/API_specification/generated/signatures.manipulation_functions.stack.html and https://data-apis.org/array-api/latest/API_specification/generated/signatures.manipulation_functions.concat.html

According to my [comments](https://github.com/numpy/numpy/blob/main/numpy/array_api/_manipulation_functions.py#L19-L20) in the code, incorrect promotion was an issue for `concat`. But I guess this is not so for `stack` (I misread the code there; `stack` does call `result_type`, but only for the error).

seberg replied:
Sounds good, might be nice to mention the fact that it only matters for `axis=None` for concat, but doesn't matter too much.  (Which makes this really a very niche usage.)

asmeurer commented on doc/source/reference/array_api.rst:
```diff
@@ -0,0 +1,795 @@
+       arrays.
+     - **Breaking**
+     - For example, ``np.array(0., dtype=float32)*np.array(0.,
+       dtype=float64)`` is ``float32``. Note that this is value-based casting
```
I can't tell what's a bug or not when it comes to value-based casting. It definitely is odd that `__pow__` is the only operator that does this.

asmeurer commented on doc/source/reference/array_api.rst:
```diff
@@ -0,0 +1,795 @@
+     - **Breaking**
+     -
+   * - ``pinv`` has an ``rtol`` keyword argument instead of ``rcond``
+     - **Compatible**
```
There was some discussion about this (also for `matrix_rank`). See https://github.com/data-apis/array-api/issues/211 and https://github.com/data-apis/array-api/issues/216.

rgommers replied:
This is breaking indeed, but as breaks go it's as small as they get.

rgommers replied:
Fine to mark as breaking here I'd say, pointing out that the default changes from `1e-15` to `None`, which means `max(M, N) * eps`.

asmeurer replied:
I marked these as compatible because the name of the keyword argument also changed.

rgommers replied:
That doesn't matter here right? A `np.linalg.pinv(x)` call would start behaving differently, so that is backwards incompatible. If you were thinking that we'd keep `rcond` _and_ `rtol`, that wouldn't quite make sense.

seberg replied:
It is an incompatible change and it could at the very least break the reproducibility of workflows (results change and I am not sure how to transition smoothly).  So it should be marked/discussed as such.  In principle it has the potential to be a large change, but I doubt it is in practice.

I would expect that the number of affected users will be rather small and that it should almost never cause _wrong_ results; so it would rather only break reproducibility.  And more often than not, it hopefully will "fix" results also.
Plus, reproducibility may well already be a game of luck for those that would be affected.

asmeurer replied:
Ah, you're right of course. The keyword rename can be done compatibly but the default change can't.

asmeurer replied:
The same reasoning also applies `matrix_rank`, and to the `stable` flag of `sort` and `argsort`.

asmeurer commented on doc/source/reference/array_api.rst:
```diff
@@ -0,0 +1,795 @@
+   * - ``cholesky`` includes an ``upper`` keyword argument.
+     - **Compatible**
+     -
+   * - ``cross`` does not broadcast its arguments.
+     - ???
+     -
+   * - ``cross`` does not allow size 2 vectors (only size 3).
+     - ???
+     -
```
I'm still unclear whether these `cross` differences are a hard requirement in the spec or not. CC @rgommers @kgryte

kgryte replied:
I've opened an issue concerning `linalg.cross` behavior: https://github.com/data-apis/array-api/issues/415. The outcome of that discussion should have bearing on potential backward compatibility concerns.

kgryte replied:
As discussed in https://github.com/data-apis/array-api/issues/415, allowing size 2 vectors is likely to never be supported in the specification given difficulties such support presents in, e.g., JIT compilation (e.g., https://github.com/numpy/numpy/issues/13718). Accordingly, we should mark this NumPy behavior as backward incompatible with the spec.

Re: broadcasting. My sense is that we may be moving toward alignment on NumPy broadcasting behavior, but this is still TBD. Atm, I don't think we need to mark NumPy's current broadcasting behavior as backward incompatible.

asmeurer committed:
Update review comments on the array API document

asmeurer committed:
Note that keyword arguments with different defaults are breaking

This applies even of the name changed, because it affects the case where no
keyword is passed.

charris commented:
Discussion has quieted down. Ready to merge?

asmeurer commented:
I'm OK with merging. There were a few open questions on some of the points (`outer`, `matrix_rank`, and `cross`), but I think it's fine to merge this now and update those later if necessary.

charris commented:
OK, let's get this in. Thanks @asmeurer ,
